### PR TITLE
Implement groupRecentActivity service

### PR DESCRIPTION
**Pull request by zenovich**

### CODE REVIEW

File: app/database/itemStore.go [hunk 1]
```diff
@@ -28,13 +28,17 @@ func (s *ItemStore) tableName() string {
 
 // Visible returns a view of the visible items for the given user
 func (s *ItemStore) Visible(user AuthUser) DB {
+	return s.KeepItemsVisibleBy(user, s.All())
+}
+
+// KeepItemsVisibleBy returns a subview of the visible items for the given user basing on the given view
+func (s *ItemStore) KeepItemsVisibleBy(user AuthUser, db DB) DB {
```
Comment (smadbe) [inline app/database/itemStore.go]: I'm not a big fan of these Keep* functions. Function chaining is so much readable for such a use. We should find a solution to support these.

File: app/api/groups/groups.go [hunk 2]
```diff
@@ -16,4 +16,5 @@ type Service struct {
 func (srv *Service) SetRoutes(router chi.Router) {
 	router.Use(auth.UserIDMiddleware(&srv.Config.Auth))
 	router.Get("/groups/", service.AppHandler(srv.getAll).ServeHTTP)
+	router.Get("/groups/recent_activity", service.AppHandler(srv.getRecentActivity).ServeHTTP)
```
Comment (smadbe) [inline app/api/groups/groups.go]: I would have expected `/groups/<group_id>/recent_activity` I think

File: app/api/groups/recent_activity.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	if err == nil {
+		query = query.Where("users_answers.bValidated = ?", validated)
+	}
+	return query
+}
+
+func (srv *Service) filterByFromSubmissionDateAndFromID(r *http.Request, query database.DB) (database.DB, error) {
+	fromID, fromIDError := service.ResolveURLQueryGetInt64Field(r, "from.id")
+	fromSubmissionDate, fromSubmissionDateError := service.ResolveURLQueryGetStringField(r, "from.submission_date")
+	if (fromIDError != nil && fromSubmissionDateError == nil) || (fromIDError == nil && fromSubmissionDateError != nil) {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/recent_activity.go]: fixed

File: app/api/groups/recent_activity.robustness.feature [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	Background:
+		Given the database has the following table 'users':
+			| ID | sLogin  | tempUser | idGroupSelf | idGroupOwned | sFirstName  | sLastName |
+			| 1  | someone | 0        | 21          | 22           | Bill        | Clinton   |
+			| 2  | user    | 0        | 11          | 12           | John        | Doe       |
+      | 3  | owner   | 0        | 23          | 24           | Jean-Michel | Blanquer  |
+		And the database has the following table 'groups_ancestors':
+			| ID | idGroupAncestor | idGroupChild | bIsSelf | iVersion |
+      | 75 | 24              | 13           | 0       | 0        |
+			| 76 | 13              | 11           | 0       | 0        |
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/api/groups/recent_activity.robustness.feature]: fixed

File: app/database/groupAncestorStore.go [hunk 5]
```diff
@@ -16,7 +16,22 @@ func (s *GroupAncestorStore) UserAncestors(user AuthUser) DB {
...
+// that are descendants of the user's owned group using AuthUser object
 func (s *GroupAncestorStore) OwnedByUser(user AuthUser) DB {
 	return s.All().Where("idGroupAncestor=?", user.OwnedGroupID())
 }
+
+// OwnedByUserID returns a composable query for getting all the groups_ancestors rows for groups
+// that are descendants of the user's owned group using ID
+func (s *GroupAncestorStore) OwnedByUserID(ownerUserID int64) DB {
+	return s.All().
+		Joins("JOIN users on users.idGroupOwned=groups_ancestors.idGroupAncestor").
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/groupAncestorStore.go]: fixed

File: app/api/groups/groups.go (hunk 2 above)
Comment (zenovich) [reply app/api/groups/groups.go]: fixed

File: app/service/utils.go [hunk 6]
```diff
@@ -55,6 +57,16 @@ func ResolveURLQueryGetBoolField(httpReq *http.Request, name string) (bool, erro
 	return strValue == "1", nil
 }
 
+// ResolveURLQueryPathInt64Field extracts a path element of type int64 from the query
+func ResolveURLQueryPathInt64Field(httpReq *http.Request, name string) (int64, error) {
```
Comment (smadbe) [inline app/service/utils.go]: It is time now to start adding unit tests each time you add a new function 😋 !

Comment (zenovich) [reply app/service/utils.go]: Sure. Why not? I love them! 👍

Comment (zenovich) [reply app/service/utils.go]: done

File: app/database/groupAncestorStore.go [hunk 7]
```diff
@@ -16,7 +16,22 @@ func (s *GroupAncestorStore) UserAncestors(user AuthUser) DB {
...
 // OwnedByUser returns a composable query for getting all the groups_ancestors rows for groups
-// that are descendants of the user's owned group
+// that are descendants of the user's owned group using AuthUser object
 func (s *GroupAncestorStore) OwnedByUser(user AuthUser) DB {
 	return s.All().Where("idGroupAncestor=?", user.OwnedGroupID())
 }
+
+// OwnedByUserID returns a composable query for getting all the groups_ancestors rows for groups
+// that are descendants of the user's owned group using ID
+func (s *GroupAncestorStore) OwnedByUserID(ownerUserID int64) DB {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/groupAncestorStore.go]: fixed

File: app/database/itemAncestorStore.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+type ItemAncestorStore struct {
+	*DataStore
+}
+
+func (s *ItemAncestorStore) tableName() string {
+	return "items_ancestors"
+}
+
+// All creates a composable query without filtering
+func (s *ItemAncestorStore) All() *ItemAncestorStore {
```
[the remark this answers is not in the document]
Comment (zenovich) [reply app/database/itemAncestorStore.go]: fixed (there are no All() anymore)

File: app/database/itemStore.go (hunk 1 above)
Comment (zenovich) [reply app/database/itemStore.go]: fixed

**Comment by zenovich** [comment]
> Reviewed from `f0c3fdd`. Not particular remarks.
> I don't know if your IDE highlights for you the lines which are covered by tests and those which are not, but there are still a lot of lines which are not covered (actually you should see it as the coverage is 59.1% in this package for now. However, it is probably out of the scope of this PR and should be done from a clean master.

I look at the coverage report here https://codecov.io/gh/France-ioi/AlgoreaBackend/tree/feat%2FgroupRecentActivity/app/

**Comment by zenovich** [comment]
@smadbe I didn't write tests methods tht were not modified/added by this PR

**Comment by smadbe** [comment]
> I look at the coverage report here https://codecov.io/gh/France-ioi/AlgoreaBackend/tree/feat%2FgroupRecentActivity/app/

It looks good to check which lines are covered as well.

Ok for me to merge everything.

**Pull request merged**